cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java

```diff
@@ -0,0 +1,509 @@
+    // Ensure all sources exist
+    for (Source source : sources) {
+      if (Source.Type.STREAM.equals(source.getType())) {
+        if (!streamExists(source.getName())) {
+          throw new IllegalArgumentException(String.format("Stream %s must exist during create of adapter: %s",
+                                                           source.getName(), adapterName));
+        }
+      } else {
+        throw new IllegalArgumentException(String.format("Unknown Source type: %s", source.getType()));
+      }
+    }
+  }
+
+  private boolean streamExists(String streamName) {
+    try {
+      return streamAdmin.exists(streamName);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  // create the required sinks for the adapters. Currently only DATASET sink type is supported.
+  private void createSinks(Set<Sink> sinks, AdapterTypeInfo adapterTypeInfo) {
+    // create sinks if it does not exist
+    for (Sink sink : sinks) {
+      if (Sink.Type.DATASET.equals(sink.getType())) {
+        String datasetName = sink.getName();
+        // add all properties that were defined in the manifest (default sink properties), override that with sink
+        // properties passed while creating the sinks.
+        DatasetProperties properties = DatasetProperties.builder()
+                                            .addAll(adapterTypeInfo.getDefaultSinkProperties())
+                                            .addAll(sink.getProperties())
+                                            .build();
+        createDataset(datasetName, properties.getProperties().get(DATASET_CLASS), properties);
+      } else {
+        throw new IllegalArgumentException(String.format("Unknown Sink type: %s", sink.getType()));
+      }
+    }
+  }
+
+  private void createDataset(String datasetName, String datasetClass, DatasetProperties properties) {
+    Preconditions.checkNotNull(datasetClass, "Dataset class cannot be null");
+    try {
+      if (!datasetFramework.hasInstance(datasetName)) {
+        datasetFramework.addInstance(datasetClass, datasetName, properties);
+        LOG.debug("Dataset instance {} created with properties: {}.", datasetName, properties);
+      } else {
+        LOG.debug("Dataset instance {} already exists; not creating a new one.", datasetName);
+      }
+    } catch (DatasetManagementException e) {
+      LOG.error("Error while creating dataset {}", datasetName, e);
+      throw new RuntimeException(e);
+    } catch (IOException e) {
+      LOG.error("Error while creating dataset {}", datasetName, e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  // Reads all the jars from the adapter directory and sets up required internal structures.
+  @VisibleForTesting
+  void registerAdapters() {
+    try {
+      File baseDir = new File(configuration.get(Constants.AppFabric.ADAPTER_DIR));
+      Collection<File> files = FileUtils.listFiles(baseDir, new String[]{"jar"}, true);
```

Comment:
I thought will be a good idea to have some folder structure if we chose to have in the future
```
/adapters/stream-conversion
/adapters/kafka-readers
```
etc. 

What do you think/?